### PR TITLE
add gas-water simulator

### PR DESCRIPTION
**Pull request by plgbrts**

### CODE REVIEW

**Comment by totto82** [comment]
Thanks for the contribution. A gas-water simulator is very handy. Is this PR sufficient to make a gas-water simulation run. Or do you plan additional PRs in opm-models?

**Comment by plgbrts** [comment]
The gas-water simulator is work in progress. This PR is not sufficient to perform a gas-water simulation and is only a first step. I created a water-gas test case and now working on opm-material for the gas-water material laws.

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Looks like a rebase would be good ;-)

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

File: flow/flow_ebos_gaswater.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+public:
+    typedef Opm::BlackOilTwoPhaseIndices<getPropValue<TypeTag, Properties::EnableSolvent>(),
+                                         getPropValue<TypeTag, Properties::EnablePolymer>(),
+                                         getPropValue<TypeTag, Properties::EnableEnergy>(),
+                                         getPropValue<TypeTag, Properties::EnableFoam>(),
+                                         getPropValue<TypeTag, Properties::EnableBrine>(),
+                                         /*PVOffset=*/0,
+                                         /*disabledCompIdx=*/FluidSystem::waterCompIdx> type;
+};
```
Comment (totto82) [inline flow/flow_ebos_gaswater.cpp:64]: This should be FluidSystem::oilCompIdx

Comment (plgbrts) [reply flow/flow_ebos_gaswater.cpp:64]: Many thanks, that resolves indeed  the  issues I mentioned to you

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by akva2** [comment]
- this merges from an upstream. always rebase on upstream.
- the merge is faulty. the new simulator is not built.

**Comment by goncalvesmachadoc** [comment]
jenkins build this please

**Comment by plgbrts** [comment]
@totto82 I made a fix for this PR can you have a look and if it is ok can it be merged? I am now working on a PR for the gas-water functionality in opm-material.

File: ebos/ebos_gaswater.cc [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+{
+private:
+    // it is unfortunately not possible to simply use 'TypeTag' here because this leads
+    // to cyclic definitions of some properties. if this happens the compiler error
+    // messages unfortunately are *really* confusing and not really helpful.
+    using FluidSystem = GetPropType<TTag::EbosTypeTag, Properties::FluidSystem>;
+
+public:
+    typedef Opm::BlackOilTwoPhaseIndices<getPropValue<TypeTag, Properties::EnableSolvent>(),
+                                         getPropValue<TypeTag, Properties::EnablePolymer>(),
```
Comment (totto82) [inline ebos/ebos_gaswater.cc:57]: You need to add getPropValue<TypeTag, Properties::EnableExtbo>() after the solvent line

**Review by totto82 — COMMENTED** [review]
Sorry for not merging this earlier. I thought it was work-in-progress. What happens if I run flow with a water gas case if this get merged? Does it produce a error message:  telling we don't support this yet somewhere in the simulator or does it crash?

File: ebos/ebos_gaswater.cc (hunk 2 above)
Comment (totto82) [reply ebos/ebos_gaswater.cc:57]: Here and also in flow part.

Comment (plgbrts) [reply ebos/ebos_gaswater.cc:58]: Thanks for reviewing. I fixed the missing arguments. There is an assert in opm-material but added a temporary error message and exit failure return in Main.hpp

**Comment by totto82** [comment]
jenkins build this please

**Comment by akva2** [comment]
The branch is still messed up - the new simulators are not built. I took the liberty of fixing it up for you, I had to fix one error (use of the wrong typetag in ebos_gaswater.cc). Fixed branch can be found as akva2/gaswater_rebased, https://github.com/akva2/opm-simulators/commits/gaswater_rebased

One simple way to use it:
```
 # Move to local git to gaswater branch
git checkout gaswater
# Add my repo as a remote
git remote add akva2 https://github.com/akva2/opm-simulators
# Fetch it
git fetch akva2
# Set your gaswater branch equal to the rebased branch
git reset --hard akva2/gaswater_rebased
# Force push it tor you repo
git push <your remote> gaswater -f
# Remove my repo again
git remote rm akva2
```

**Comment by atgeirr** [comment]
> Fixed branch can be found...

Can't we just close this and open a new PR to that branch?

**Comment by akva2** [comment]
of course we can, i just assumed original author wants it merged from his repo, not mine.

**Comment by atgeirr** [comment]
> of course we can, i just assumed original author wants it merged from his repo, not mine.

As long as the commits are correctly attributed by author, I think that does not matter, but @plgbrts what do you think?

**Comment by akva2** [comment]
jenkins build this please

**Comment by plgbrts** [comment]
> > of course we can, i just assumed original author wants it merged from his repo, not mine.
> 
> As long as the commits are correctly attributed by author, I think that does not matter, but @plgbrts what do you think?

I followed the instructions of Arne

**Comment by plgbrts** [comment]
> The branch is still messed up - the new simulators are not built. I took the liberty of fixing it up for you, I had to fix one error (use of the wrong typetag in ebos_gaswater.cc). Fixed branch can be found as akva2/gaswater_rebased, https://github.com/akva2/opm-simulators/commits/gaswater_rebased
> 
> One simple way to use it:
> 
> ```
>  # Move to local git to gaswater branch
> git checkout gaswater
> # Add my repo as a remote
> git remote add akva2 https://github.com/akva2/opm-simulators
> # Fetch it
> git fetch akva2
> # Set your gaswater branch equal to the rebased branch
> git reset --hard akva2/gaswater_rebased
> # Force push it tor you repo
> git push <your remote> gaswater -f
> # Remove my repo again
> git remote rm akva2
> ```

Thanks for helping me out

**Comment by totto82** [comment]
> The branch is still messed up

@akva2 Please merge if your issue is fixed.

**Pull request merged**